### PR TITLE
build: Upgrade PRQL version

### PR DESCRIPTION
**Pull request by max-sixty**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Upgrade PRQL to 0.9.3

---

Hopefully this is OK. I couldn't build ClickHouse locally, but `cargo build` within `ClickHouse/rust/prql` works well.

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit cc5ec9e6345fbbf154536b3a7a95f3ea9b6452d1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53060/cc5ec9e6345fbbf154536b3a7a95f3ea9b6452d1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by alexey-milovidov — APPROVED** [review]

**Comment by azat** [comment]
>outdated PRs fail to build after such changes with the following error

Which in particular?
It should update the docker image.

**Comment by Felixoid** [comment]
The updated image should work everywhere, even in outdated PRs.

I can't provide you a link, it's in the private repo. But the scheme, where an updated image breaks builds in other branches, won't work

**Comment by Felixoid** [comment]
To reproduce, here's the commands:

```
docker pull clickhouse/binary-builder:latest
git checkout e5d3e348ce65e9a5bcce3d3658f53b85e5716765
./docker/packager/packager --package-type=binary --output-dir "./$(git rev-parse --show-cdup)/build_results"
```

**Comment by Felixoid** [comment]
It won't, I didn't try pretending it will. I just invite you there to discuss, how we could make the sources available, but won't be so strict.

Maybe, we could consider other schemes? The first wild idea is pre-downloading them in `docker/packager/packager` and mount into the container. Then they could be stored as formerly ccache in https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/ccache_utils.py

The state now is stable enough for some days, but another update will be a disaster again.

**Comment by azat** [comment]
> Maybe, we could consider other schemes? The first wild idea is pre-downloading them in docker/packager/packager and mount into the container. Then they could be stored as formerly ccache in [master/tests/ci/ccache_utils.py](https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/ccache_utils.py?rgh-link-date=2023-08-07T15%3A02%3A03Z)

This is a good idea I guess, since a) we can keep rust folder where it is now and b) ~it should has less time of invalid cache~.

Actually thinking about this one more time, I don't see why that build failed, github actions uses MERGE_HEAD, so they always use recent changes (regardless of was there merge with upstream or not), and so once the docker image is new, they should use them and nothing should fail.

From your private logs I see:

>Aug 07 11:42:28 error: failed to select a version for the requirement `prql-compiler = "^0.8.1"`
>Aug 07 11:42:28 candidate versions found which didn't match: 0.9.3

Which means that it uses old version of sources.

**Comment by Felixoid** [comment]
> github actions uses MERGE_HEAD
which is a branching point. Once it's created, it won't change for at least the current workflow running. Because GH actions preserve the merge commit at the push time.

Trying to understand how it works and how we could use it is one of the deepest rabbit holes I've ever seen, so let's not go there :-j 

> Which means that it uses old version of sources.

And our build system must consider it. If an updated build image won't work for older commits, we should change it ASAP. Any PR should work, and updating Rust dependencies must be a smooth and trivial operation. Now it requires ASAP updating anything using building image, including old releases.

> This is a good idea I guess, since a) we can keep rust folder where it is now

We can mount sources there. But wouldn't it act as it's now? It's actually possible to update them on the fly :thinking: 

Can we update and clean unnecessary dependencies at the build time?

**Comment by azat** [comment]
>Trying to understand how it works and how we could use it is one of the deepest rabbit holes I've ever seen, so let's not go there :-j

I haven't looked deeply enough into actions/checkout (or your analog), but AFAIR it should use MERGE_HEAD by default, but maybe there are some differences when there are conflicts, which means that you cannot have MERGE_HEAD.

>And our build system must consider it. If an updated build image won't work for older commits, we should change it ASAP. Any PR should work, and updating Rust dependencies must be a smooth and trivial operation. Now it requires ASAP updating anything using building image, including old releases.

Not sure that I understand this part, like I already wrote, it should use new version of sources always, regardless of how ancient the PR is.

The only exception I guess is the PRs that created for release branches, well, then it should use different docker image tags, and AFAIR it does so.

>We can mount sources there. But wouldn't it act as it's now? It's actually possible to update them on the fly thinking

Yeah, I don't see how this will change anything either, since you need the build of not the new sources by the PR's sources.

>Can we update and clean unnecessary dependencies at the build time?

Well the whole point of #49601 was to avoid requiring internet connection for the build (at least for Rust). Of course simple update of vendored crates will help, but this does not sounds right (I'm not saying that it is the same as not using vendored crates at all, but still).

**Comment by max-sixty** [comment]
Thanks for merging! Lmk anything I should do to get https://github.com/ClickHouse/clickhouse-docs/pull/1417 merged too!

**Comment by Felixoid** [comment]
> Not sure that I understand this part, like I already wrote, it should use new version of sources always, regardless of how ancient the PR is.

It never does. The GH actions always use the same [ref](https://github.com/ClickHouse/ClickHouse/actions/runs/5767112236/job/15640411438#step:3:68), and it does not [change](https://github.com/ClickHouse/ClickHouse/actions/runs/5767112236/job/15641757148#step:3:68) across the run. It's necessary to have the same code for the whole running. I bet even across the [workflows](https://github.com/ClickHouse/ClickHouse/actions/runs/5767112197/job/15636261906#step:2:60)

What you are referencing is the `refs/pull/$NUMBER/merge` commit. I think, it's changed at the push time, but not 100%. So, every running PR is broken. Yes, it's slightly better, but it's still broken.

> Well the whole point of #49601 was to avoid requiring internet connection for the build (at least for Rust). Of course simple update of vendored crates will help, but this does not sounds right (I'm not saying that it is the same as not using vendored crates at all, but still).

Valid. I'd rather a bit rephrase it to 'to avoid dependency on a 3rd party sites'. I think, we could use the S3 buckets to store the prepared dependencies. It looks optimal to me. And we could name it as a hash from `Cargo.lock` file, so it will be static and won't require frequent uploads

**Comment by azat** [comment]
>What you are referencing is the refs/pull/$NUMBER/merge commit.

Yes.

>I think, it's changed at the push time, but not 100%.

I have the same guess.

>So, every already_running_PR is broken. Yes, it's slightly better, but it's still broken.

Indeed it should be like that. And if I understood you correctly in S3 you are suggesting to store more then one version of this cache?

**Comment by azat** [comment]
Just to make it clear: I'm with you here, this comment not about my objections, just to discuss.

> 100% worth it.

Actually I'm more then happy that you are trying to solve this! :)

> we have old release branches, that should be able using the latest build image w/o rush backporting changed dependencies there.

But shouldn't release branches uses per-release docker images? That said that I don't expect any problems there with Rust.

> In other words, devs will have a suddenly broken CI. Without additional attention, they would push something in the evening, come in the morning, and find nothing finished because the image was updated. It's a very frustrating situation, especially with a moderate delay between changes being pushed and the CI is completely finished.

Yes, but this should happens only if there was concurrent image update. And AFAIR there was some problems even when clang got updated.

Oh, and one more question about the implementation cache in S3 will not include the PR number right? Only hash of the `Cargo.lock`? (so that this cache will be available always, even before PR with Rust changes got merged, so that if it was merged in between of some CI started for some PR it will always has this cache)

>We already spend too much computing time on unnecessary PR updates.

I definitely agree with you here

>The state that it's good is very controversial IMHO

I'm just talking about PRs that are not updated for months, but about constant rebases to make CI green - I find them evil (if nothing had been changed in between at least)

**Comment by Felixoid** [comment]
> But shouldn't release branches uses per-release docker images? That said that I don't expect any problems there with Rust.

Only public releases. The private repo always uses the `:latest` images everywhere. There's a plan to fix it, but it's like that from the real beginning. I think all the time you are speaking about public releases, so the issue wasn't clear enough.

> Yes, but this should happens only if there was concurrent image update. And AFAIR there was some problems even when clang got updated.

Clang update is a quite rare case. Rust dependencies are updated at least once a month for some time already, and it causes additional friction each time. So it's time to put some effort and fix it.

> Only hash of the Cargo.lock?

Right. The dependencies' "snapshot" will be global across all the builds. No need to make it a per-PR thing.
